Checkpoint para el Alumno
# **Modulo 3: Procesamiento de Datos con Python**

**Introducción:**

El artículo 134 de la Constitución Política de los Estados Unidos Mexicanos establece que los recursos económicos de que dispongan la Federación, las entidades federativas,
los Municipios y las demarcaciones territoriales de la Ciudad de México, se administrarán con eficiencia,
eficacia, economía, transparencia y honradez para satisfacer los objetivos a los que estén destinados.

Así mismo establece que las adquisiciones, arrendamientos y enajenaciones de todo tipo de bienes, prestación de servicios de
cualquier naturaleza y la contratación de obra que realicen, se adjudicarán o llevarán a cabo a través de
licitaciones públicas mediante convocatoria pública para que libremente se presenten proposiciones
solventes en sobre cerrado, que será abierto públicamente, a fin de asegurar al Estado las mejores
condiciones disponibles en cuanto a precio, calidad, financiamiento, oportunidad y demás circunstancias
pertinentes.

Conociendo esto, en el siguiente proyecto haremos el análisis de una base de datos del Gobierno de México de los contratos derivados de los procedimientos de contratación reportados en CompraNet por las unidades compradoras en 2020.  

**Importamos Librerias**

Como punto inicial de nuestro cuaderno, importamos librerias y montamos el drive que estaremos utilizando durante el desarrollo del proyecto.

In [1]:
import pandas as pd
import numpy as np

**Importamos Datos**

En este punto necesitamos importar los datos con los cuales estareamos trabajando durante el proyecto.

In [2]:
contratos = pd.read_csv('/content/drive/MyDrive/Checkpoint Modulo 3/Contratos2020.csv',index_col=0)

contratos.head()

/tmp/ipykernel_5722/3782819030.py:1: DtypeWarning: Columns (6,20,42,43) have mixed types. Specify dtype option on import or set low_memory=False.
  contratos = pd.read_csv('/content/drive/MyDrive/Checkpoint Modulo 3/Contratos2020.csv',index_col=0)


,Siglas de la Institución,Institución,Clave de la UC,Nombre de la UC,Responsable de la UC,Código del expediente,Referencia del expediente,Clave CUCOP,Título del expediente,Plantilla del expediente,...,Clave de cartera SHCP,Folio en el RUPC,RFC,Proveedor o contratista,Estratificación de la empresa,Clave del país de la empresa,RFC verificado en el SAT,Crédito externo,Organismo financiero,Dirección del anuncio
Orden de gobierno,,,,,,,,,,,,,,,,,,,,,
APF,DICONSA,"Diconsa, S.A. de C.V.",008VSS998,DICONSA-Dirección Comercial #008VSS998,José Juan Velazquez Barbosa,2163712,511/32120,2380,CAFE SOLUBLE,05. Adjudicación Directa LAASSP,...,NaN,NaN,NaN,"MARCAS NESTLE, S.A. DE C.V.",NaN,MX,0,NaN,NaN,https://compranet.hacienda.gob.mx/esop/guest/g...
GM,GTO,_Gobierno Municipal del Estado de Guanajuato,811023986,GTO-Pénjamo-Tesorería Municipal #811023986,Salvador Puebla Pérez,2181836,MPE-CASCOS BALÍSTICOS-FORTASEG20-LR13/2020,2830,ADQUISICIÓN DE CASCOS BALÍSTICOS NIVEL III-A,03. Invitación a Cuando Menos Tres Personas LA...,...,NaN,NaN,MTC130813PN2,MGR TEXTILES Y CONFECCIONES SA DE CV,Pequeña,MX,1,NaN,NaN,https://compranet.hacienda.gob.mx/esop/guest/g...
APF,SEDATU,"Secretaría de Desarrollo Agrario, Territorial ...",015000999,SEDATU-Dirección General de Recursos Materiale...,Ada Gricelda Bonifaz Villar,2173823,SEDATU-856-110-2020,3390,PRESTACION DE SERVICIOS,05. Adjudicación Directa LAASSP,...,NaN,NaN,NaN,MAYDE SOLÍS MÉNDEZ,NaN,MX,0,NaN,NaN,https://compranet.hacienda.gob.mx/esop/guest/g...
APF,SEDENA,Secretaría de la Defensa Nacional,007000999,SEDENA-Subdirección de Adquisiciones #007000999,Enrique Hernández Cisneros,2177310,AA-007000999-E681-2020,3180,contratacion del servicio de mensajeria,05. Adjudicación Directa LAASSP,...,NaN,8319.0,EME880309SK5,ESTAFETA MEXICANA SA DE CV,No MIPYME,MX,1,NaN,NaN,https://compranet.hacienda.gob.mx/esop/guest/g...
APF,SEDENA,Secretaría de la Defensa Nacional,007000999,SEDENA-Subdirección de Adquisiciones #007000999,Enrique Hernández Cisneros,2177310,AA-007000999-E681-2020,3180,contratacion del servicio de mensajeria,05. Adjudicación Directa LAASSP,...,NaN,8319.0,EME880309SK5,ESTAFETA MEXICANA SA DE CV,No MIPYME,MX,1,NaN,NaN,https://compranet.hacienda.gob.mx/esop/guest/g...


**Limpieza de Datos**

Para poder proceder con la limpieza de nuestros datos, primero hacemos un analisis para saber el tamaño de nuestro DS, así como el tipo de datos con los cuales estaremos trabajando.

In [3]:
contratos.shape

(159726, 44)

In [4]:
contratos.dtypes

,0
Siglas de la Institución,object
Institución,object
Clave de la UC,object
Nombre de la UC,object
Responsable de la UC,object
Código del expediente,object
Referencia del expediente,object
Clave CUCOP,object
Título del expediente,object
Plantilla del expediente,object


Utilizando la función info del DF, podremos ver mas información de cada columna.

In [5]:
contratos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 159726 entries, APF to APF
Data columns (total 44 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Siglas de la Institución       159726 non-null  object 
 1   Institución                    159726 non-null  object 
 2   Clave de la UC                 159726 non-null  object 
 3   Nombre de la UC                159726 non-null  object 
 4   Responsable de la UC           159726 non-null  object 
 5   Código del expediente          159726 non-null  object 
 6   Referencia del expediente      150396 non-null  object 
 7   Clave CUCOP                    159713 non-null  object 
 8   Título del expediente          159726 non-null  object 
 9   Plantilla del expediente       159586 non-null  object 
 10  Fundamento legal               135875 non-null  object 
 11  Número del procedimiento       159586 non-null  object 
 12  Fecha de fallo                 33458

Después de visualizar el conteo de los valores nulos en nuestra base de datos, haremos un conteo de los valores N/A presentes en nuestra información.

In [6]:
contratos.isna()

,Siglas de la Institución,Institución,Clave de la UC,Nombre de la UC,Responsable de la UC,Código del expediente,Referencia del expediente,Clave CUCOP,Título del expediente,Plantilla del expediente,...,Clave de cartera SHCP,Folio en el RUPC,RFC,Proveedor o contratista,Estratificación de la empresa,Clave del país de la empresa,RFC verificado en el SAT,Crédito externo,Organismo financiero,Dirección del anuncio
Orden de gobierno,,,,,,,,,,,,,,,,,,,,,
APF,False,False,False,False,False,False,False,False,False,False,...,True,True,True,False,True,False,False,True,True,False
GM,False,False,False,False,False,False,False,False,False,False,...,True,True,False,False,False,False,False,True,True,False
APF,False,False,False,False,False,False,False,False,False,False,...,True,True,True,False,True,False,False,True,True,False
APF,False,False,False,False,False,False,False,False,False,False,...,True,False,False,False,False,False,False,True,True,False
APF,False,False,False,False,False,False,False,False,False,False,...,True,False,False,False,False,False,False,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
APF,False,False,False,False,False,False,True,False,False,False,...,False,True,False,False,False,False,False,False,False,False
APF,False,False,False,False,False,False,True,False,False,False,...,False,True,False,False,False,False,False,False,False,False
APF,False,False,False,False,False,False,True,False,False,False,...,False,True,False,False,False,False,False,False,False,False


In [7]:
contratos.isna().sum(axis=0)

,0
Siglas de la Institución,0
Institución,0
Clave de la UC,0
Nombre de la UC,0
Responsable de la UC,0
Código del expediente,0
Referencia del expediente,9330
Clave CUCOP,13
Título del expediente,0
Plantilla del expediente,140


Para poder trabajar con un DF más reducido y que contenga únicamente los valores de mayor interés para el análisis propuesto, vamos a eliminar todas las columnas que contienen valores N/A.

In [8]:
contratos_clean = contratos.dropna(axis=1, how='any')

In [9]:
contratos_clean.isna()

,Siglas de la Institución,Institución,Clave de la UC,Nombre de la UC,Responsable de la UC,Código del expediente,Título del expediente,Tipo de procedimiento,Código del contrato,Título del contrato,Fecha de inicio del contrato,Importe del contrato,Moneda del contrato,Estatus del contrato,Proveedor o contratista,Clave del país de la empresa,RFC verificado en el SAT
Orden de gobierno,,,,,,,,,,,,,,,,,
APF,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
GM,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
APF,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
APF,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
APF,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
APF,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
APF,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
APF,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [10]:
contratos_clean.isna().sum(axis=0)

,0
Siglas de la Institución,0
Institución,0
Clave de la UC,0
Nombre de la UC,0
Responsable de la UC,0
Código del expediente,0
Título del expediente,0
Tipo de procedimiento,0
Código del contrato,0
Título del contrato,0


Nuestro nuevo DF quedó conformado con un total de 17 columnas y 159,726 filas.

In [11]:
contratos_clean.shape

(159726, 17)

In [12]:
contratos_clean.head()

,Siglas de la Institución,Institución,Clave de la UC,Nombre de la UC,Responsable de la UC,Código del expediente,Título del expediente,Tipo de procedimiento,Código del contrato,Título del contrato,Fecha de inicio del contrato,Importe del contrato,Moneda del contrato,Estatus del contrato,Proveedor o contratista,Clave del país de la empresa,RFC verificado en el SAT
Orden de gobierno,,,,,,,,,,,,,,,,,
APF,DICONSA,"Diconsa, S.A. de C.V.",008VSS998,DICONSA-Dirección Comercial #008VSS998,José Juan Velazquez Barbosa,2163712,CAFE SOLUBLE,ADJUDICACIÓN DIRECTA,2378065,CAFE SOLUBLE,2020-09-11,42704.50,MXN,Expirado,"MARCAS NESTLE, S.A. DE C.V.",MX,0
GM,GTO,_Gobierno Municipal del Estado de Guanajuato,811023986,GTO-Pénjamo-Tesorería Municipal #811023986,Salvador Puebla Pérez,2181836,ADQUISICIÓN DE CASCOS BALÍSTICOS NIVEL III-A,INVITACIÓN A CUANDO MENOS 3 PERSONAS,2412672,ADQUISICIÓN DE CASCOS BALÍSTICOS NIVEL III-A,2020-11-17,409655.04,MXN,Expirado,MGR TEXTILES Y CONFECCIONES SA DE CV,MX,1
APF,SEDATU,"Secretaría de Desarrollo Agrario, Territorial ...",015000999,SEDATU-Dirección General de Recursos Materiale...,Ada Gricelda Bonifaz Villar,2173823,PRESTACION DE SERVICIOS,ADJUDICACIÓN DIRECTA,2388648,PRESTACION DE SERVICIOS,2020-10-09,65981.33,MXN,Expirado,MAYDE SOLÍS MÉNDEZ,MX,0
APF,SEDENA,Secretaría de la Defensa Nacional,007000999,SEDENA-Subdirección de Adquisiciones #007000999,Enrique Hernández Cisneros,2177310,contratacion del servicio de mensajeria,ADJUDICACIÓN DIRECTA,2554392,contratacion del servicio de mensajeria,2020-12-03,800000.00,MXN,Expirado,ESTAFETA MEXICANA SA DE CV,MX,1
APF,SEDENA,Secretaría de la Defensa Nacional,007000999,SEDENA-Subdirección de Adquisiciones #007000999,Enrique Hernández Cisneros,2177310,contratacion del servicio de mensajeria,ADJUDICACIÓN DIRECTA,2628175,contratación del servicio de mensajería,2020-12-03,800000.00,MXN,Expirado,ESTAFETA MEXICANA SA DE CV,MX,1


Guardamos el resultado para tener el documento disponible para otras consultas:

In [13]:
contratos_clean.to_csv('/content/drive/MyDrive/Checkpoint Modulo 3/Contratos2020_clean.csv')

**Realizamos un análisis exploratorio de los datos**

Ya con nuestro DF limpio podemos hacer ciertas consultas para contestar algunas preguntas.

In [14]:
contratos_clean.groupby('Tipo de procedimiento').size()

,0
Tipo de procedimiento,
ADJUDICACIÓN DIRECTA,125057
CONTRATO ENTRE ENTES PUBLICOS,881
INVITACIÓN A CUANDO MENOS 3 PERSONAS,10825
LICITACIÓN PÚBLICA,18963
OTRAS CONTRATACIONES,3953
PROYECTO DE CONVOCATORIA,47


In [15]:
contratos_clean.groupby('Tipo de procedimiento')['Siglas de la Institución'].value_counts()

Tipo de procedimiento     Siglas de la Institución
ADJUDICACIÓN DIRECTA      IMSS                        37782
                          DICONSA                     19473
                          ISSSTE                       7866
                          OnceTV                       2716
                          CINVESTAV                    1968
                                                      ...  
PROYECTO DE CONVOCATORIA  SICT                           10
                          CDMX                            3
                          HRAEB                           1
                          ISSSTE                          1
                          SEMAR                           1
Name: count, Length: 1122, dtype: int64

In [16]:
contratos_clean.groupby(['Tipo de procedimiento', 'Siglas de la Institución'])['Importe del contrato'].sum()

Tipo de procedimiento     Siglas de la Institución
ADJUDICACIÓN DIRECTA      AEFCM                       1.424515e+08
                          AEM                         8.512542e+06
                          AFAC                        2.995835e+06
                          AGN                         3.840537e+06
                          AGROASEMEX                  1.573872e+07
                                                          ...     
PROYECTO DE CONVOCATORIA  HRAEB                       5.030204e+06
                          IMSS                        5.420775e+06
                          ISSSTE                      3.358610e+07
                          SEMAR                       8.824000e+04
                          SICT                        4.349606e+08
Name: Importe del contrato, Length: 1122, dtype: float64

In [17]:
contratos_clean.groupby('Tipo de procedimiento')['Importe del contrato'].sum()

,Importe del contrato
Tipo de procedimiento,
ADJUDICACIÓN DIRECTA,1.977724e+11
CONTRATO ENTRE ENTES PUBLICOS,9.857796e+09
INVITACIÓN A CUANDO MENOS 3 PERSONAS,2.057045e+10
LICITACIÓN PÚBLICA,2.141518e+11
OTRAS CONTRATACIONES,5.805677e+10
PROYECTO DE CONVOCATORIA,6.542016e+08
